python/GafferSceneUI/_SceneViewInspector.py

```diff
@@ -0,0 +1,959 @@
+#  - acquireEdit [callable] will return the plug that performs the edit, creating it if necessary.
+#  - editType [_EditType] the type of edit that (will) be facilitated by the plug.
+#  - inEditScope [Bool] True if the node affecting the edit is a child of an EditScope.
+#  - hasEdit [Bool] True if the plugs for the edit already exist, and in the case of tweaks,
+#       if they are enabled.
+_EditInfo = namedtuple( "_EditInfo", [ "acquireEdit", "editType", "inEditScope", "hasEdit" ] )
+
+## \todo Figure out how this relates to the inspectors in the SceneInspector.
+## \see _EditInfo for the meaning of the various accessors.
+class _ParameterInspector( object ) :
+
+	def __init__( self, attributeHistory, parameter, editScope ) :
+
+		self.__parameter = parameter
+		self.__errorMessage = None
+		self.__warningMessages = []
+
+		shader = attributeHistory.attributeValue.outputShader()
+		self.__value = shader.parameters.get( parameter )
+		if self.__value is not None :
+			if hasattr( self.__value, "value" ) :
+				self.__value = self.__value.value
+			self.__editInfo = self.__findEdit( attributeHistory, editScope )
+		else :
+			self.__editInfo = None
+
+	def value( self ) :
+
+		return self.__value
+
+	def editable( self ) :
+
+		return self.__editInfo is not None
+
+	def acquireEdit( self ) :
+
+		return self.__editInfo.acquireEdit() if self.__editInfo else None
+
+	def editType( self ) :
+
+		return self.__editInfo.editType  if self.__editInfo else None
+
+	def hasEdit( self ) :
+
+		return self.__editInfo.hasEdit if self.__editInfo else None
+
+	def inEditScope( self ) :
+
+		return self.__editInfo.inEditScope if self.__editInfo else False
+
+	def warningMessage( self ) :
+
+		return "/n".join( self.__warningMessages )
+
+	def errorMessage( self ) :
+
+		return self.__errorMessage if self.__errorMessage is not None else ""
+
+	## \todo This is a copy of TransformTool::spreadsheetAwareSource, and is
+	# eerily similar to Dispatcher::computedSource and others. This needs
+	# factoring out into some general contextAwareSource method that can be
+	# shared by all these.
+	def __spreadsheetAwareSource( self, plug ) :
+
+		plug = plug.source()
+
+		spreadsheet = plug.node()
+		if not isinstance( spreadsheet, Gaffer.Spreadsheet ) :
+			return plug
+
+		if not spreadsheet["out"].isAncestorOf( plug ) :
+			return pluge
+
+		plug = spreadsheet.activeInPlug( plug )
+		if plug.ancestor( Gaffer.Spreadsheet.RowPlug ).isSame( spreadsheet["rows"].defaultRow() ) :
+			return None
+
+		return plug.source()
+
+	def __editFromSceneNode( self, attributeHistory ) :
+
+		node = attributeHistory.scene.node()
+		if not isinstance( node, ( GafferScene.ShaderTweaks, GafferScene.Light, GafferScene.ShaderAssignment ) ) :
+			return None
+
+		if attributeHistory.scene != node["out"] :
+			return None
+
+		if isinstance( node, GafferScene.Light ) :
+
+			plug = self.__spreadsheetAwareSource( node["parameters"][ self.__parameter ] )
```

Comment:
I think we need to scope `attributeHistory.context()` here, in case it affects the outcome of `spreadsheetAwareSource()`. We might be better off to do the scoping once at outside the node-type conditionals, or possible even at the point we call `__editFromSceneNode()`.

Reply:
Improved logic [here](https://github.com/GafferHQ/gaffer/pull/3832/commits/c5fd1610f5e6e61763ea25870cad5942531c4d43).